### PR TITLE
Send ticks to the encoder when dropping frames

### PR DESCRIPTION
**Pull request by fiban-havok**
Mitigates https://github.com/microsoft/MixedReality-WebRTC/issues/74.

Merged some small refactorings too which don't change behavior.

### CODE REVIEW

File: third_party/winuwp_h264/H264Encoder/H264Encoder.cc [hunk 1]
```diff
@@ -346,10 +356,19 @@ int WinUWPH264EncoderImpl::Encode(
...
+      sample = FromVideoFrame(frame);
     }
-    sample = FromVideoFrame(frame);
+  }
+
+  if (!sample) {
+    // Drop the frame. Send a tick to keep the encoder going.
+    lastFrameDropped_ = true;
+    auto timestampHns = GetFrameTimestampHns(frame);
+    ON_SUCCEEDED(sinkWriter_->SendStreamTick(streamIndex_, timestampHns));
```
Comment (jehumb-havok) [inline third_party/winuwp_h264/H264Encoder/H264Encoder.cc:372]: The [docs of `IMFSinkWriter::SendStreamTick()`](https://docs.microsoft.com/en-us/windows/win32/api/mfreadwrite/nf-mfreadwrite-imfsinkwriter-sendstreamtick#remarks) says that we should set the [`MFSampleExtension_Discontinuity` attribute](https://docs.microsoft.com/en-us/windows/win32/medfound/mfsampleextension-discontinuity-attribute) on the first sample after the discontinuity. Do we want to implement that too with the current change?

File: third_party/winuwp_h264/H264Encoder/H264Encoder.cc [hunk 2]
```diff
@@ -312,6 +318,10 @@ ComPtr<IMFSample> WinUWPH264EncoderImpl::FromVideoFrame(const VideoFrame& frame)
   return sample;
 }
 
+LONGLONG WinUWPH264EncoderImpl::GetFrameTimestampHns(const VideoFrame& frame) const {
+  return ((frame.timestamp() - startTime_) / 90) * 1000 * 10;
```
Comment (jehumb-havok) [inline third_party/winuwp_h264/H264Encoder/H264Encoder.cc]: Can you please add a comment and reference https://tools.ietf.org/html/rfc6184#page-11 which states that:
> A 90 kHz clock rate MUST be used.

This is the actual reason why we have this magical "90" number here. The actual formula is:
```
timestamp_100ns = timestamp_90kHz / {90'000 kHz} * {10'000'000 hns/sec}
```

File: third_party/winuwp_h264/H264Encoder/H264Encoder.cc [hunk 3]
```diff
@@ -535,12 +554,12 @@ int WinUWPH264EncoderImpl::SetRates(
...
 
     EncodedImageCallback* tempCallback = encodedCompleteCallback_;
-    Release();
+    ReleaseWriter();
     {
       rtc::CritScope lock(&callbackCrit_);
       encodedCompleteCallback_ = tempCallback;
     }
-    InitEncoderWithSettings(&codec_);
+    InitEncoderWithSettings();
```
Comment (jehumb-havok) [inline third_party/winuwp_h264/H264Encoder/H264Encoder.cc]: Can we remove the discrepancy between "Init**Encoder**WithSettings" and "Release**Writer**"? I believe those two functions work in pair, so should have similar names, no? Can we just rename to `InitWriter()` instead?

File: third_party/winuwp_h264/H264Encoder/H264Encoder.h [hunk 4]
```diff
@@ -56,16 +56,16 @@ class WinUWPH264EncoderImpl : public VideoEncoder, public IH264EncodingCallback
 
  private:
   ComPtr<IMFSample> FromVideoFrame(const VideoFrame& frame);
-  int InitEncoderWithSettings(const VideoCodec* codec_settings);
+  int InitEncoderWithSettings();
+  int ReleaseWriter();
+  LONGLONG GetFrameTimestampHns(const VideoFrame& frame) const;
```
Comment (jehumb-havok) [inline third_party/winuwp_h264/H264Encoder/H264Encoder.h:61]: Document that "Hns" means "hundreds of nanoseconds"? (the standard Media Foundation timestamp unit)

**Review by jehumb-havok — CHANGES_REQUESTED** [review]

File: third_party/winuwp_h264/H264Encoder/H264Encoder.cc (hunk 1 above)
Comment (fiban-havok) [reply third_party/winuwp_h264/H264Encoder/H264Encoder.cc:372]: The code is already doing this, see line 314. It's checking lastFrameDropped_ even if that was never set to false before this change 😄

File: third_party/winuwp_h264/H264Encoder/H264Encoder.h (hunk 4 above)
Comment (fiban-havok) [reply third_party/winuwp_h264/H264Encoder/H264Encoder.h:61]: TBH I didn't know what it meant, I just copied the existing names 😄 Will do.

File: third_party/winuwp_h264/H264Encoder/H264Encoder.cc (hunk 1 above)
Comment (jehumb-havok) [reply third_party/winuwp_h264/H264Encoder/H264Encoder.cc:372]: I'll put a coin in the "Better GitHub code review" jar. 😒

File: third_party/winuwp_h264/H264Encoder/H264Encoder.h (hunk 4 above)
Comment (jehumb-havok) [reply third_party/winuwp_h264/H264Encoder/H264Encoder.h:61]: I suspected as much. I would never have guessed without reading the RFC of RTP while looking for something else.

**Review by jehumb-havok — APPROVED** [review]

**Review by LoadLibrary — APPROVED** [review]

**Pull request merged**